### PR TITLE
Added two more tests to cover config update by creating an image w a new tag 

### PR DESCRIPTION
**Pull request by hzhao-github**
local run success and 5 test cases passed:

[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3,210.989 s - in oracle.kubernetes.operator.ItModelInImageConfigUpdate
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Build Tools 1.0 .................................... SUCCESS [  1.189 s]
[INFO] weblogic-kubernetes-operator 3.0.0 ................. SUCCESS [  2.683 s]
[INFO] operator-swagger 3.0.0 ............................. SUCCESS [  7.367 s]
[INFO] json-schema 3.0.0 .................................. SUCCESS [  9.130 s]
[INFO] jsonschema-maven-plugin Maven Mojo 3.0.0 ........... SUCCESS [  8.207 s]
[INFO] operator-runtime 3.0.0 ............................. SUCCESS [ 59.556 s]
[INFO] operator-integration-tests 3.0.0 ................... SUCCESS [54:39 min]
[INFO] installation-tests 3.0.0 ........................... SUCCESS [  2.161 s]
[INFO] Project Reports 3.0.0 .............................. SUCCESS [  0.096 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  56:09 min
[INFO] Finished at: 2020-04-02T16:21:37-07:00
[INFO] ------------------------------------------------------------------------

### CODE REVIEW

**Comment by markxnelson** [comment]
have you run this on jenkins yet?

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 1]
```diff
@@ -109,7 +109,7 @@ public static void staticUnPrepare() throws Exception {
    * @throws Exception if domain creation, config update or test veriofication fails
    */
   @Test
-  public void testMiiConfigUpdateNonExistJdbc() throws Exception {
+  public void testMiiConfigUpdateNonJdbcCm() throws Exception {
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:112]: tests should not throw exceptions

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 2]
```diff
@@ -153,6 +153,62 @@ public void testMiiConfigUpdateNonExistJdbc() throws Exception {
...
+   * Create a domain without a JDBC DS using model in image and having configmap
+   * in the domain.yaml before deploying the domain. After deploying the domain crd,
+   * create a new image with diff tag name and model files that define a JDBC DataSource
+   * and update the domain crd to change image name to reload the model,
+   * generate new config and initiate a rolling restart.
+   *
+   * @throws Exception if domain creation, config update or test veriofication fails
+   */
+  @Test
+  public void testMiiConfigUpdateNonJdbcImage() throws Exception {
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:166]: tests should not throw exceptions
if you think there may be an exception - write an assert to catch it

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 3]
```diff
@@ -153,6 +153,62 @@ public void testMiiConfigUpdateNonExistJdbc() throws Exception {
...
+      // verify that JDBC DS is created by checking JDBC DS name and read timeout
+      LoggerHelper.getLocal().log(Level.INFO, "Verify that JDBC DS is created");
+      Set<String> jdbcResourcesToVerify = new HashSet<String>();
+      jdbcResourcesToVerify.add("datasource.name.1=" + dsName);
+      jdbcResourcesToVerify.add("datasource.readTimeout.1=" + readTimeout_2);
+
+      final String destDir = getResultDir() + "/samples/model-in-image-update";
+      verifyJdbcResources(jdbcResourcesToVerify, destDir);
+
+      testCompletedSuccessfully = true;
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:202]: is it just me, or is there not a single assertion in this whole test?

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 4]
```diff
@@ -153,6 +153,62 @@ public void testMiiConfigUpdateNonExistJdbc() throws Exception {
...
   }
 
+  /**
+   * Create a domain without a JDBC DS using model in image and having configmap
+   * in the domain.yaml before deploying the domain. After deploying the domain crd,
+   * create a new image with diff tag name and model files that define a JDBC DataSource
+   * and update the domain crd to change image name to reload the model,
+   * generate new config and initiate a rolling restart.
+   *
+   * @throws Exception if domain creation, config update or test veriofication fails
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:163]: spelling

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:163]: fixed changed to verification

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 3 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:202]: assertion is inside of verifyJdbcResources(jdbcResourcesToVerify, destDir);

  private void verifyJdbcResources(Set<String> jdbcResourcesSet, String destDir)  {
    // verify JDBC DS props via WLST on server pod
    String jdbcResources = getJdbcResources(destDir);

    Iterator<String> iterator = jdbcResourcesSet.iterator();
    while (iterator.hasNext()) {
      String prop = iterator.next();
      Assertions.assertTrue(jdbcResources.contains(prop), prop + " is not found");
      LoggerHelper.getLocal().log(Level.INFO, prop + " exists");
    }
  }

Comment (markxnelson) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:202]: it feels like there should be more than that, i think we should be checking each operation as we go, to make sure it is behaving as we expect.  your javadoc reads like there are several steps - we should check as we go to make sure they all work, please

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 5]
```diff
@@ -218,6 +274,62 @@ public void testMiiConfigUpdateExistJdbc() throws Exception {
...
+   * Create a domain with a JDBC DS using model in image and having configmap
+   * in the domain.yaml before deploying the domain. After deploying the domain crd,
+   * create a new image with diff tag name and model files that define a JDBC DataSource
+   * and update the domain crd to change image name to reload the model,
+   * generate new config and initiate a rolling restart.
+   *
+   * @throws Exception if domain creation, config update or test veriofication fails
+   */
+  @Test
+  public void testMiiConfigUpdateJdbcImage() throws Exception {
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:287]: i have the same concern here, it seems like there are several steps and there should be assertions after each step to make sure it is working as desired. thanks

File: integration-tests/src/test/java/oracle/kubernetes/operator/MiiBaseTest.java [hunk 6]
```diff
@@ -76,15 +77,109 @@ public Domain createMiiDomainWithConfigMap(String domainUidPrefix,
...
+      .append("/miiWorkDir/models/archive.zip")
+      .append(" --wdtModelOnly ")
+      .append(" --wdtDomainType ")
+      .append(WdtDomainType.WLS.geWdtDomainType());
+
+    // patching the domain
+    LoggerHelper.getLocal().log(Level.INFO, "Command to create domain image: " + createDomainImageScriptCmd);
+    ExecResult result = ExecCommand.exec(createDomainImageScriptCmd.toString(), true, additionalEnvMap);
+    if (result.exitValue() != 0) {
+      throw new RuntimeException(
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/MiiBaseTest.java:134]: this is good - we are checking the outcome, and printing out a nice descriptive message, but i think an assertion that it does not throw an exception would be better style.

File: integration-tests/src/test/java/oracle/kubernetes/operator/MiiBaseTest.java [hunk 7]
```diff
@@ -76,15 +77,109 @@ public Domain createMiiDomainWithConfigMap(String domainUidPrefix,
...
+        .append(" -p='[{\"op\": \"replace\", \"path\": \"/spec/image\", \"value\": \"'")
+        .append(imageName)
+        .append("'\" }]'");
+
+    // patching the domain
+    LoggerHelper.getLocal().log(Level.INFO, "Command to patch domain: " + patchDomainCmd);
+    ExecResult result = TestUtils.exec(patchDomainCmd.toString());
+    LoggerHelper.getLocal().log(Level.INFO, "Domain patch result: " + result.stdout());
+
+    // verify the domain restarted
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/MiiBaseTest.java:170]: logging is great, thanks!
i also would love to see an assertion in here, and a descriptive error please

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 2 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:166]: removed "throws Exception" from all tests

File: integration-tests/src/test/java/oracle/kubernetes/operator/MiiBaseTest.java (hunk 6 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/MiiBaseTest.java:134]: yea. I have changed it to:

catch (Exception ex) {
      StringBuffer errorMsg = new StringBuffer("FAILURE: command: ");
      errorMsg
          .append(createDomainImageScriptCmd)
          .append(" failed, returned ")
          .append(result.stdout())
          .append("\n")
          .append(result.stderr());

      Assertions.fail(errorMsg + "\n " + ex.getMessage());
    }

File: integration-tests/src/test/java/oracle/kubernetes/operator/MiiBaseTest.java (hunk 7 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/MiiBaseTest.java:170]: yea. I have changed it to:

      StringBuffer errorMsg = new StringBuffer("FAILURE: command: ");
      errorMsg
          .append(patchDomainCmd)
          .append(" failed, returned ")
          .append(result.stdout())
          .append("\n")
          .append(result.stderr());

      Assertions.fail(errorMsg + "\n " + ex.getMessage());

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 1 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:112]: fixed

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 5 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:287]: ok. Let me move the assertions to right after the operation is done to make it look clearer

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java]: Mark, I made changes regarding this. I added Assertions after each step, some of them are added in the helper methods, others added in the test methods themselves. Could you please take one more look see if you are ok with the changes.

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 8]
```diff
@@ -105,48 +103,137 @@ public static void staticUnPrepare() throws Exception {
...
+      // update domain yaml with restartVersion,
+      // apply the domain yaml and verify domain restarted successfully
       modifyDomainYamlWithRestartVersion(domain, domainNS);
 
-      // verify the test result by checking updated config file on server pod
-      verifyJdbcUpdate();
+      // verify that JNDI name exists by checking updated config file on server pod
+      String jdbcDsStr = getJndiName(domain);
+      Assertions.assertTrue(jdbcDsStr.contains("<jndi-name>" + jndiName + "</jndi-name>"),
+          "JDBC DS doesn't update");
```
Comment (markxnelson) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:139]: do you mean it wasn't updated?

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java]: fixed.

**Comment by markxnelson** [comment]
@ddsharpe @rjeberhard I am happy to merge this

**Comment by rjeberhard** [comment]
@hzhao-github, remove the title "WIP" when you are ready to merge this PR.

**Comment by hzhao-github** [comment]
Still waiting for a pass on External Jenkins. Seems External Jenkins broken, nothing passed for days

**Pull request merged**